proto/merge.go

```diff
@@ -73,103 +221,142 @@ func mergedFileDescriptors(debug bool) (*descriptorpb.FileDescriptorSet, error)
-			// mismatch, then we do nothing, and take the protoregistry file
-			// descriptor as the correct one.
-			if debug && !protov2.Equal(protodesc.ToFileDescriptorProto(protoregFd), fd) {
-				diff := cmp.Diff(protodesc.ToFileDescriptorProto(protoregFd), fd, protocmp.Transform())
-				diffErr = append(diffErr, fmt.Sprintf("Mismatch in %s:\n%s", *fd.Name, diff))
+			// Otherwise it was a not found error, so add it.
+			// At this point we can't validate.
+			p.fdCh <- fd
+			continue
+		}
+
+		if validate {
+			fdp := protodesc.ToFileDescriptorProto(protoregFd)
+			if !protov2.Equal(fdp, fd) {
+				diff := cmp.Diff(fdp, fd, protocmp.Transform())
+				ec.DiffCh <- fmt.Sprintf("Mismatch in %s:\n%s", *fd.Name, diff)
 			}
 		}
 	}
+}
 
-	if debug {
-		errStr := new(bytes.Buffer)
-		if len(checkImportErr) > 0 {
-			fmt.Fprintf(errStr, "Got %d file descriptor import path errors:\n\t%s\n", len(checkImportErr), strings.Join(checkImportErr, "\n\t"))
-		}
-		if len(diffErr) > 0 {
-			fmt.Fprintf(errStr, "Got %d file descriptor mismatches. Make sure gogoproto and protoregistry use the same .proto files. '-' lines are from protoregistry, '+' lines from gogo's registry.\n\n\t%s\n", len(diffErr), strings.Join(diffErr, "\n\t"))
-		}
-		if errStr.Len() > 0 {
-			return nil, fmt.Errorf(errStr.String())
-		}
+// collectFDs runs in its own goroutine, exhausing p.fdCh to populate p.fds,
+// and then sorting p.fds in-place.
+func (p *descriptorProcessor) collectFDs() {
+	defer p.fdWG.Done()
+
+	for fd := range p.fdCh {
+		p.fds = append(p.fds, fd)
 	}
 
-	slices.SortFunc(fds.File, func(x, y *descriptorpb.FileDescriptorProto) bool {
+	slices.SortFunc(p.fds, func(x, y *descriptorpb.FileDescriptorProto) bool {
 		return *x.Name < *y.Name
 	})
-
-	return fds, nil
 }
 
-// MergedRegistry returns a *protoregistry.Files that acts as a single registry
-// which contains all the file descriptors registered with both gogoproto and
-// protoregistry (the latter taking precendence if there's a mismatch).
-func MergedRegistry() (*protoregistry.Files, error) {
-	fds, err := MergedFileDescriptors()
-	if err != nil {
-		return nil, err
+// mergedFileDescriptors coordinates an instance of a descriptorProcessor
+// and a descriptorErrorCollector to concurrently merge the file descriptors in globalFiles and appFiles,
+// into a new *descriptorpb.FileDescriptorSet.
+//
+// If validate is true, do extra work to validate that import paths are properly formed
+// and that "duplicated" file descriptors across globalFiles and appFiles
+// are indeed identical, returning an error if either of those conditions are invalidated.
+func mergedFileDescriptors(globalFiles *protoregistry.Files, appFiles map[string][]byte, validate bool) (*descriptorpb.FileDescriptorSet, error) {
+	// GOMAXPROCS is the number of CPU cores available, by default.
+	// Respect that setting as the number of CPU-bound goroutines,
+	// and for channel sizes.
+	nProcs := runtime.GOMAXPROCS(0)
```

Comment:
limit with CPU-bound is a good idea, but might need rename to [mergeFileDescriptors](https://github.com/cosmos/gogoproto/blob/9d031c6a9d21756d032dbdf8221474ce3674c490/proto/merge.go#L190)

Reply:
Good catch, thank you. Comments corrected in cf9db7151b1b6aa0b1963fcb574f7514e20fbc7e.